README.md

```diff
@@ -273,7 +278,103 @@ should mention how to instantiate those variables/files, etc.
+and the expected results. Now, every time you make a change to your code, you need to go through every single item on that list and check it.
+
+That doesn’t sound like much fun, does it?
+
+This is where automated testing comes in. Automated testing is the execution of your test plan (the parts of your application you want to test, 
+the order in which you want to test them, and the expected responses) by a script instead of a human. Software testing involves the execution of a 
+software component or system component to evaluate one or more properties of interes. In general, these properties indicate the extent to which the component
+ or system under test:
+
+* meets the requirements that guided its design and development,
+* responds correctly to all kinds of inputs,
+* performs its functions within an acceptable time,
+* is sufficiently usable,
+* can be installed and run in its intended environments, and
+* achieves the general result its stakeholders desire.
+
+#### Test driven development
+Test-driven development (TDD) is a software development process that relies on the repetition of a very short development cycle: requirements are turned into 
+very specific test cases, then the code is improved so that the tests pass. In simple terms, 
+test cases for each functionality are created and tested first and if the test fails then the new code is written in order to pass the test 
+and making code simple and bug-free.
+
+
+
+Useful links for writing tests in our most common  languages and frameworks: 
+* [Python](https://realpython.com/python-testing/)
+* [React](https://reactjs.org/docs/testing-recipes.html)
+* [Javascript](https://jestjs.io/)
+* [R](https://r-pkgs.org/tests.html)
+
+## Continuous integration and delivery
+
+The CI/CD pipeline is one of the best practices for devops teams to implement, 
+for delivering code changes more frequently and reliably. It is also an agile methodology 
+best practice, as it enables software development teams to focus on meeting business requirements, 
+and code quality because deployment steps are automated.
+
+### Continuous integration
+Developers practicing continuous integration merge their changes back to the main branch as often as possible. 
+The developer's changes are validated by creating a build and running automated tests against the build. 
+By doing so, you avoid the integration hell that usually happens when people wait for release day to merge
+ their changes into the release branch.
+
+Continuous integration puts a great emphasis on testing automation to check that the application is not broken 
+whenever new commits are integrated into the main branch.
+
+#### What you need (cost)
+Your team will need to write automated tests for each new feature, improvement or bug fix.
+You need a continuous integration server that can monitor the main repository and run the tests automatically for every new commits pushed.
```

Comment:
s/server/service/

Possibly add something about the cost of time taken after push, and whatever to differentiate tests when a change is "ready" vs still worked on.

Reply:
Not sure what you mean here. Can you add more details?

Reply:
Even if more testing is better, having time consuming tests may be frustrating and disturb the flow.

One example now is where we have tests taking roughly 10-15 minutes to run. I want to see those run successfully before I push, so after every time (I think) I'm done, there's a 15 minutes waiting period. As its long enough to not wait for, I need to remember to check on the tests and commit/push if everything's ready (if it's not, there will be another such cycle).

So, in some cases, it may be useful having a smaller set of test cases that are considered acceptable as a minimum threshold for what may be on a feature branch and a larger set of tests (typically labelled integration tests) that should pass before merge.

(And in some cases, that's not suitable, but it's worth noting the problem I think.)

Reply:
Mmm. Interesting, but I am not sure something is worth to mention in the main text. I do not know how to integrate it here at least. If you have a paragraph that works please send it, or some useful links.